localization/localization_util/include/localization_util/smart_pose_buffer.hpp

```diff
@@ -23,33 +23,41 @@
 
 #include <deque>
 
-class PoseArrayInterpolator
+class SmartPoseBuffer
 {
 private:
   using PoseWithCovarianceStamped = geometry_msgs::msg::PoseWithCovarianceStamped;
 
 public:
-  PoseArrayInterpolator(
-    rclcpp::Node * node, const rclcpp::Time & target_ros_time,
-    const std::deque<PoseWithCovarianceStamped::ConstSharedPtr> & pose_msg_ptr_array,
-    const double & pose_timeout_sec, const double & pose_distance_tolerance_meters);
+  struct InterpolateResult
+  {
+    PoseWithCovarianceStamped old_pose;
+    PoseWithCovarianceStamped new_pose;
+    PoseWithCovarianceStamped interpolated_pose;
+  };
 
-  PoseArrayInterpolator(
-    rclcpp::Node * node, const rclcpp::Time & target_ros_time,
-    const std::deque<PoseWithCovarianceStamped::ConstSharedPtr> & pose_msg_ptr_array);
+  SmartPoseBuffer() = delete;
+  SmartPoseBuffer(
+    const rclcpp::Logger & logger, const double & pose_timeout_sec,
+    const double & pose_distance_tolerance_meters);
 
-  PoseWithCovarianceStamped get_current_pose();
-  PoseWithCovarianceStamped get_old_pose();
-  PoseWithCovarianceStamped get_new_pose();
-  [[nodiscard]] bool is_success() const;
+  std::optional<InterpolateResult> interpolate(const rclcpp::Time & target_ros_time);
+
+  void push_back(const PoseWithCovarianceStamped::ConstSharedPtr & pose_msg_ptr);
+
+  void pop_old(const rclcpp::Time & target_ros_time);
+
+  void clear();
+
+  bool detect_time_jump_to_past(const rclcpp::Time & target_ros_time);
 
 private:
   rclcpp::Logger logger_;
-  rclcpp::Clock clock_;
-  const PoseWithCovarianceStamped::SharedPtr current_pose_ptr_;
-  PoseWithCovarianceStamped::SharedPtr old_pose_ptr_;
-  PoseWithCovarianceStamped::SharedPtr new_pose_ptr_;
-  bool success_;
+  std::deque<PoseWithCovarianceStamped::ConstSharedPtr> pose_buffer_;
+  std::mutex mutex_;
```

Comment:
Please specify that what this `mutex_` is for. :pray:
e.g. `// This mutex is for pose_buffer_`

Reply:
Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5713/commits/548e105068e68e2b5affac392dd4b749cac525ea